### ISSUE TITLE
GPCP and SST plots are not shown when comparing to obs

### ISSUE THREAD

**Issue by tangq** [E3SM-Project/acme_diags#161]
I ran the diagnostics yesterday, but the GPCP and SST figures are missing (http://portal.nersc.gov/project/acme/tang30/FC5AV1C-04P2.ne30.ne30_ne30.syrah/e3sm_diags/0002-0005_vs_obs/viewer/latitude-longitude/index.html)

The log files show the following errors.

ncopen: filename "/global/project/projectdirs/acme/acme_diags/obs_for_acme_diags/GPCP_v2.2_ANN_climo.nc": NetCDF: HDF error
Error in acme_diags.driver.lat_lon_driver
Traceback (most recent call last):
  File "/global/homes/t/tang30/.conda/envs/acme_diags_env/lib/python2.7/site-packages/acme_diags/acme_diags_driver.py", line 78, in run_diag
    single_result = module.run_diag(parameters)
  File "/global/homes/t/tang30/.conda/envs/acme_diags_env/lib/python2.7/site-packages/acme_diags/driver/lat_lon_driver.py", line 75, in run_diag
    f_obs = cdms2.open(filename2)
  File "/global/homes/t/tang30/.conda/envs/acme_diags_env/lib/python2.7/site-packages/cdms2/dataset.py", line 371, in openDataset
    file1 = CdmsFile(path, "r")
  File "/global/homes/t/tang30/.conda/envs/acme_diags_env/lib/python2.7/site-packages/cdms2/dataset.py", line 1098, in __init__
    raise CDMSError('Cannot open file %s (%s)' % (path, err))
CDMSError: Cannot open file /global/project/projectdirs/acme/acme_diags/obs_for_acme_diags/GPCP_v2.2_ANN_climo.nc (NetCDF: HDF error)

**Comment by tangq** [comment]
It looks like a NetCDF file version issue. The netCDF 3 files does not have the error.

(acme_diags_env) [tang30@edison03 logs]$ ncdump -k /global/project/projectdirs/acme/acme_diags/obs_for_acme_diags/GPCP_v2.2_ANN_climo.nc
netCDF-4 classic model
(acme_diags_env) [tang30@edison03 logs]$ ncdump -k /global/project/projectdirs/acme/acme_diags/obs_for_acme_diags/ERA-Interim_11_1980_2016_climo.nc
classic

**Comment by chengzhuzhang** [comment]
@tangq I recreated your env, using `v1.3.2` of `e3sm_diags` and it works fine. **However, I used a different model file, since I did not have permission to yours.**

Here's the `myparam.py` file that I used:
```python
reference_data_path = '/global/project/projectdirs/acme/acme_diags/obs_for_acme_diags/'
test_data_path = '/global/project/projectdirs/acme/acme_diags/test_model_data_for_acme_diags/'
test_name = '20161118.beta0.FC5COSP.ne30_ne30.edison'

sets = ["lat_lon"]
backend = 'mpl'
results_dir = 'lat_lon_demo'  # name of folder where all results will be stored
```

Using the following `mydiags_GPCP.cfg` file, we just ran the diags for `GPCP` and it worked:
```
[#]
sets = ["lat_lon"]
case_id = "GPCP_v2.2"
variables = ["PRECT"]
ref_name = "GPCP_v2.2"
reference_name = "GPCP (yrs1979-2014)"
seasons = ["ANN", "DJF", "MAM", "JJA", "SON"]
regions = ["global"]
test_colormap = "WhiteBlueGreenYellowRed.rgb"
reference_colormap = "WhiteBlueGreenYellowRed.rgb"
diff_colormap = "BrBG"
contour_levels = [0.5, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 12, 13, 14, 15, 16]
diff_levels = [-5, -4, -3, -2, -1, -0.5, 0.5, 1, 2, 3, 4, 5]
```

You run it like so:
```
acme_diags -p myparams.py -d mydiags_GPCP.cfg
```

@tangq Did you happen to try to use `conda update` or `conda install`? Because this is known to cause errors like this. When upgrading to a new version of `e3sm_diags`, please create a new env from the `yml` files.

@zshaheen can help you more with these env issues.

**Comment by tangq** [comment]
I reinstalled env from scratch. But it seems that it still uses the file named "acme_diags_env.yml". I see there is another file call "e3sm_diags_env.yml" (https://raw.githubusercontent.com/E3SM-Project/acme_diags/master/conda/e3sm_diags_env.yml).

I will give it another try with e3sm_diags_env.yml

**Comment by tangq** [comment]
@chengzhuzhang , did you run your job on the edison login node or computing node?

I reinstalled my env with the v1.3.3 e3sm_diags. The GPCP plots were created successfully on the login node, but not on the computing node.

Can you try your test with an interactive session with the following commands to see if you can reproduce the error?

salloc -N 1 -q debug
source activate your_env
acme_diags -p myparams.py -d mydiags_GPCP.cfg

**Comment by chengzhuzhang** [comment]
Good point! It took long to generate the interactive session so I only did it on login node. Let me try..

**Comment by chengzhuzhang** [comment]
@tangq I also just got this error right now. We are investigating it, it's quite strange. Both the env in the login node and the interactive sessions are the same, I just checked. Also, `$PATH` is the same as well.

I will create another env, using the version that @golaz uses `v1.2.X`. That seems to work on the interactive nodes.

We will keep you posted on the progress. Thank you again for using the software!

**Comment by chengzhuzhang** [comment]
More trouble shooting:
Tried to run 'v1.2.1', 'v1.3.0', 'v1.3.1' by creating the environment individually. This HDF issue happens since 'v1.3.1'. Following are the individual ymls from v1.3.0 and v1.3.1

[env_131.txt](https://github.com/E3SM-Project/acme_diags/files/1953696/env_131.txt)
[env_130.txt](https://github.com/E3SM-Project/acme_diags/files/1953697/env_130.txt)

The major change is CDAT version  from 1.3.0 and 1.3.1, since there is really not too much diagnostics-wise change, I think for now, @tangq and @golaz may use 1.3.0 until we fix this issue...

**Comment by tangq** [comment]
Note that the login node does not have this problem. But I don't know what changes between the login and computing nodes.

**Comment by golaz** [comment]
This is rather odd indeed. Since we know that it has to do with netcdf4 files, it could be related to the hdf libraries needed for netcdf4. I wonder whether it picks up system hdf libraries in one instance instead of the ones installed with conda.

**Comment by golaz** [comment]
I believe I found a solution or at least workaround for this problem. The original error message (before the CDMS error reported by @tangq above) is:

> H5FDsec2.c line 940 in H5FD_sec2_lock(): unable to lock file, errno = 524, error message = 'Unknown error 524'

A quick search of that specific error message with Google leads to a workaround that works for me:

`export HDF5_USE_FILE_LOCKING=FALSE`

**Longer explanation**

From https://support.hdfgroup.org/ftp/HDF5/current/src/unpacked/release_docs/RELEASE.txt :

>       The file locking calls used in HDF5 1.10.0 (including patch 1)
>       will fail when the underlying file system does not support file
>       locking or where locks have been disabled. To disable all file
>       locking operations, an environment variable named
>       HDF5_USE_FILE_LOCKING can be set to the five-character string
>       'FALSE'. This does not fundamentally change HDF5 library
>       operation (aside from initial file open/create, SWMR is lock-free),
>       but users will have to be more careful about opening files
>       to avoid problematic access patterns (i.e.: multiple writers)
>       that the file locking was designed to prevent.

Now, comparing versions of HDF5 in e3sm_diags environments from @chengzhuzhang above:

```
env_131.txt:
  - hdf5=1.10.1=h9caa474_1
```

```
env_130.txt:
  - hdf5=1.8.18=1
```

I think that's the explanation. Since 1.3.1, we are using hdf5 1.10.1 which requires HDF5_USE_FILE_LOCKING=FALSE "_when the underlying file system does not support file
locking or where locks have been disabled_". This may also explains why it was working on the Edison interactive nodes but not the compute nodes.

@zshaheen : can you help with this?
1. Independently verify that this works for you.
2. Implement a fix: either modify the conda environment to automatically set the HDF5_USE_FILE_LOCKING environment variable to FALSE (if that's possible), or do it within the python code.

**Comment by oksanaguba** [comment]
I had somewhat similar issue: https://github.com/E3SM-Project/acme_diags/issues/164 .

There is more traffic now in acme_diagnostics, would it make sense to direct github activity to a slack channel?

**Comment by golaz** [comment]
Thanks, @oksanaguba. I did see your issue earlier but somehow I did not make the connection with the netCDF4 error; my mistake.

I'm not an active user of slack (maybe I should be), so I can't comment about that.

**Comment by zshaheen** [comment]
Thank you @golaz, @oksanaguba, and others. I can report now that this issue is fixed after testing it myself. I implemented the fix in 526f0871a9a7c29534aadc03182b79b947cafb84.

I tested with the `py` and `cfg` file @chengzhuzhang gave [here](https://github.com/E3SM-Project/acme_diags/issues/161#issuecomment-384798859). I got the error before the fix, and after it, everything was okay. I ran it like so, just how @tangq explained:
```
salloc -N 1 -q debug
source activate YOUR_ENV
e3sm_diags -p issue_161.py -d issue_161.cfg
```

For anyone who wants this fix, please follow the instructions titled ["Environment for development"](https://e3sm-project.github.io/acme_diags/docs/html/install.html#environment-for-development) to create an env with the latest code on master.

Thank you all for your patience and input. It's greatly appreciated.

**Comment by zshaheen** [comment]
For the cdms people (@dnadeau4 and @doutriaux1) I'm not sure if it's appropriate to force the environmental variable `HDF5_USE_FILE_LOCKING = FALSE` by default with cdms. However, since `hdf5` is a dependency of cdms, you guys should know about this.
